### PR TITLE
Port changes over from web branch

### PR DESCRIPTION
**Pull request by NicoG96**
* Added model as field for TagModel
* Tweak to train.py to save/update model
* Remove unused imports
* Code cleanup and language conventions

### CODE REVIEW

File: js/classes/tagmodel.js [hunk 1]
```diff
@@ -6,10 +6,10 @@ class TagModel {
     this.openDocs = [];
     this.currentCategory = null;
     this.categories = [];
+    this.currentModel = null;
```
Comment (tuj52136) [inline js/classes/tagmodel.js]: This probably shouldn't default to null, since it causes an error trying to join the file paths.  Maybe set to "", or set to a default "untitled" or something, so that the path join works, unless you named it somewhere else

File: js/backend.js [hunk 2]
```diff
@@ -36,8 +36,9 @@ $('#trainNew').on('click', function () {
             if (data.filePaths[0]) {
                 console.log("Changing paths to: '" + data.filePaths[0] + "'");
                 modelPath = path.join(data.filePaths[0].goodPath(), modelName);
+                tagModel.currentModel = modelPath;
                 $('#trainCurrent').show();
-                $('#trainName').text(modelPath.truncStart(30, true)).show();
+                $('#trainName').text(modelPath.truncStart(30, true));
```
[the remark this answers is not in the document]
Comment (NicoG96) [reply js/backend.js]: woops

File: js/backend.js [hunk 3]
```diff
@@ -35,10 +32,9 @@ $('#trainNew').on('click', function () {
             // on close get file path
             if (data.filePaths[0]) {
                 console.log("Changing paths to: '" + data.filePaths[0] + "'");
-                modelPath = path.join(data.filePaths[0].goodPath(), modelName);
```
Comment (tuj52136) [inline js/backend.js]: yeah you right.  I added this thinking 'load new' was specifying a dir to house a new model, but if this is for 'load existing' then we're golden, and we can just add a 'create blank model' button using the saveDialog instead. In either case we don't need the name

File: py/train.py [hunk 4]
```diff
@@ -33,7 +33,8 @@ def data_converting(raw_data):
             final_data.append((d['text'],temp))
     return final_data, label_set
 
-def main(output_dir, data_path, n_iter, model=None):
+
+def main(data_path, n_iter, model):
```
Comment (tuj52136) [inline py/train.py]: something's not working quite right here.  We're still passing 4 args from the parse. Also these are named wrong still from the web version.  data_path should be raw_data now, since it's just the actual json content.  model should be model_path

Comment (NicoG96) [reply py/train.py]: I'm only seeing 3 arguments being passed. Also output_dir and model would be the same value so I don't think it's necessary to have both

Comment (tuj52136) [reply py/train.py]: in train.py, lines 88-118 is where the arg parsing happens, that's where the issue is, specifically 118-121 i think. It might be easier to keep 4 args: 3 args could work, but train should also create a model if it's not already there, so you would have an output path to save it, but you might not have an actual model there yet.  I think it would error in that case, so you'd have to handle it,  probably changing something line 42.

Comment (NicoG96) [reply py/train.py]: Ahh I see the parser thing now.  I think it would make more sense to keep the main function reserved for training and then separate creating a new model into a function of its own, though. Thoughts?

Comment (tuj52136) [reply py/train.py]: the purist side of me thinks yes separate func, since it seems like a whole separate path, but since the entire model creation is done in 2 lines it would probably be overkill

Comment (NicoG96) [reply py/train.py]: Right, but if we were to click a 'create new model' button we wouldn't necessarily want to train it right away, just create it.  I think that's reason enough to just make a 2 line create_model() function. @tmarkn be our arbiter and settle the dispute i dont care either way just 2 different approaches

Comment (NicoG96) [reply py/train.py]: Also separating logical blocks of code into discrete components is good practice as it makes unit tests a million times easier

Comment (tuj52136) [reply py/train.py]: That's true, but wait, are you talking about 1. a separate function in the python code, 2. a separate script in the python code, or 3. a separate function in the js code

Comment (NicoG96) [reply py/train.py]: Yea probably all 3

Comment (tuj52136) [reply py/train.py]: now you've got me all for it, but i'm trying to do it and it's not as easy as i thought.  the model creation isn't so bad, but the setup for a new model is a little different. i don't know if there's a way to tell if it's been set up or not beforehand (at the moment it checks if you gave it a model to load, but in our case we always give it a model, it just might be blank).  the optimizer we give it in the update method i think attaches it to it, but eff me if i can figure out how to put it in there.  
long story short, so far creating without training is an effing pain

Comment (tuj52136) [reply py/train.py]: tried one in features/create_script, you can see if it works for you

**Comment by NicoG96** [comment]
closed for changes
